libraries/standard/http/src/http_client.c

```diff
@@ -29,6 +34,91 @@ static HTTPStatus_t _sendHttpBody( const HTTPTransportInterface_t * pTransport,
+ * @param pValue The header value to write.
+ * @param valueLen The byte length of the header field value.
+ *
+ * @return #HTTP_SUCCESS if successful. If there was insufficient memory in the
+ * application buffer, then #HTTP_INSUFFICIENT_MEMORY is returned.
+ */
+static HTTPStatus_t _addHeader( HTTPRequestHeaders_t * pRequestHeaders,
+                                const char * pField,
+                                size_t fieldLen,
+                                const char * pValue,
+                                size_t valueLen );
+
+/*-----------------------------------------------------------*/
+
+static HTTPStatus_t _addHeader( HTTPRequestHeaders_t * pRequestHeaders,
+                                const char * pField,
+                                size_t fieldLen,
+                                const char * pValue,
+                                size_t valueLen )
+{
+    HTTPStatus_t returnStatus = HTTP_SUCCESS;
+    uint8_t * pBufferCur = pRequestHeaders->pBuffer + pRequestHeaders->headersLen;
+    size_t toAddLen = 0;
+    size_t initHeadersLen = pRequestHeaders->headersLen;
+    uint8_t hasTrailingLine = 0u;
+
+    /* Backtrack before trailing "\r\n" (HTTP header end) if it's already written.
+     * Note that this method also writes trailing "\r\n" before returning. */
+    if( strncmp( ( char * ) pBufferCur - ( 2 * HTTP_HEADER_LINE_SEPARATOR_LEN ),
+                 "\r\n\r\n", 2 * HTTP_HEADER_LINE_SEPARATOR_LEN ) == 0 )
+    {
+        /* Set this flag to backtrack in case of HTTP_INSUFFICIENT_MEMORY. */
+        hasTrailingLine = 1u;
+        pBufferCur -= HTTP_HEADER_LINE_SEPARATOR_LEN;
+    }
+
+    /* Check if there is enough space in buffer for additional header. */
+    toAddLen = fieldLen + HTTP_HEADER_FIELD_SEPARATOR_LEN + valueLen +
+               HTTP_HEADER_LINE_SEPARATOR_LEN +
+               HTTP_HEADER_LINE_SEPARATOR_LEN;
+
+    /* If we have enough room for the new header line, then write it to the header buffer. */
+    if( ( initHeadersLen + toAddLen ) <= pRequestHeaders->bufferLen )
+    {
+        if( hasTrailingLine == 1u )
+        {
+            pRequestHeaders->headersLen = initHeadersLen - HTTP_HEADER_LINE_SEPARATOR_LEN;
+        }
+
+        /* Write "Field: Value \r\n\r\n" to headers. */
+        pBufferCur += snprintf( ( char * ) pBufferCur,
+                                toAddLen,
+                                "%.*s%s%.*s%s",
+                                ( int ) fieldLen, pField,
+                                HTTP_HEADER_FIELD_SEPARATOR,
+                                ( int ) valueLen, pValue,
+                                HTTP_HEADER_LINE_SEPARATOR );
+
+        /* HTTP_HEADER_LINE_SEPARATOR cannot be written above because snprintf
+         * writes an extra null byte at the end. */
+        memcpy( pBufferCur, HTTP_HEADER_LINE_SEPARATOR, HTTP_HEADER_LINE_SEPARATOR_LEN );
+        pRequestHeaders->headersLen += toAddLen;
```

Comment:
With my suggestions above you can add the toAddLen to initHeadersLen and then assign it to pRequestHeaders->headersLen.